From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Sleigh-InSPECtor <sleighinspector@outlook.com>
Date: Fri, 31 May 2024 01:46:46 +0930
Subject: [PATCH] 6606: AArch32: vmov had incorrect vmask & missing position
 shift of vector element being copied

AArch32: fixed vmov

* Incorrect vmask and missing position shift of vector element being copied.
---
 Ghidra/Processors/ARM/data/languages/ARMneon.sinc | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/Ghidra/Processors/ARM/data/languages/ARMneon.sinc b/Ghidra/Processors/ARM/data/languages/ARMneon.sinc
index ff9a4de602..a0a60b4fa0 100644
--- a/Ghidra/Processors/ARM/data/languages/ARMneon.sinc
+++ b/Ghidra/Processors/ARM/data/languages/ARMneon.sinc
@@ -4087,8 +4087,8 @@ dNvmovIndex: Dn^"["^vmovIndex^"]"   is Dn & vmovIndex     { }
                                    ($(TMODE_E) &         thv_c2327=0x1c & thv_c2222=1 & thv_c2020=0 & thv_c0811=11 & thv_c0404=1 & thv_c0003=0 ) ) & COND & Dn & VRd & vmovIndex & dNvmovIndex
 {
 	el:1 = VRd(0);
-	vmask:8 = 0xf << (vmovIndex*8);
-	Dn = (Dn & ~vmask) | (zext(el) & vmask);
+	vmask:8 = 0xff << (vmovIndex*8);
+	Dn = (Dn & ~vmask) | (zext(el) << (vmovIndex*8));
 	#VectorSetElement(VRd,Dn,vmovIndex);
 }
 
@@ -4096,8 +4096,8 @@ dNvmovIndex: Dn^"["^vmovIndex^"]"   is Dn & vmovIndex     { }
                                        ($(TMODE_E) &         thv_c2327=0x1c & thv_c2222=0 & thv_c2020=0 & thv_c0811=11 & thv_c0505=1 & thv_c0404=1 & thv_c0003=0 ) ) & COND & Dn & VRd & vmovIndex & dNvmovIndex
 {
 	el:2 = VRd(0);
-	vmask:8 = 0xff << (vmovIndex*16);
-	Dn = (Dn & ~vmask) | (zext(el) & vmask);
+	vmask:8 = 0xffff << (vmovIndex*16);
+	Dn = (Dn & ~vmask) | (zext(el) << (vmovIndex*16));
 	#VectorSetElement(VRd,Dn,vmovIndex,vmovSize);
 }
 
@@ -4106,8 +4106,8 @@ dNvmovIndex: Dn^"["^vmovIndex^"]"   is Dn & vmovIndex     { }
                                        ($(TMODE_E) &         thv_c2327=0x1c & thv_c2222=0 & thv_c2020=0 & thv_c0811=11 & thv_c0506=0 & thv_c0404=1 & thv_c0003=0 ) ) & COND & Dn & VRd & vmovIndex & dNvmovIndex
 {
 	el:4 = VRd;
-	vmask:8 = 0xffff << (vmovIndex*32);
-	Dn = (Dn & ~vmask) | (zext(el) & vmask);
+	vmask:8 = 0xffffffff << (vmovIndex*32);
+	Dn = (Dn & ~vmask) | (zext(el) << (vmovIndex*32));
 	#VectorSetElement(VRd,Dn,vmovIndex,vmovSize);
 }
 
-- 
2.45.1

